simulation/conveyor/model.py

```diff
@@ -0,0 +1,325 @@
+      conveyor speed (T might be negative, moving objects backwards)
+    - calculate time T until earliest event of object reaching a checkpoint
+      (or the end of the conveyor), and also return those checkpoint and object
+    """
+
+    def __init__(self, env: Environment, length: float, checkpoints: list[dict[str, int | Section]], model_id: int):
+        assert length > 0, "Conveyor length <= 0!"
+
+        checkpoints = sorted(checkpoints, key=lambda p: p['position'])
+        if len(checkpoints) > 0:
+            assert checkpoints[0]['position'] >= 0, "Checkpoints with position < 0!"
+            assert checkpoints[-1]['position'] < length, "Checkpoint with position >= conveyor length!"
+            for i in range(len(checkpoints) - 1):
+                assert checkpoints[i]['position'] < checkpoints[i + 1]['position'], \
+                    "Checkpoints with equal positions!"
+
+        # constants
+        self._env = env
+        self._model_id = model_id
+        self._checkpoints = checkpoints
+        self._checkpoint_positions = {cp['node']: cp['position'] for cp in checkpoints}
+        self._length = length
+
+        # variables
+        self._speed = 1
+        self._objects = {}
+        self._object_positions = []
+
+        self._state = 'static'
+        self._resume_time = 0
+        self._resolved_events = set()
+
+    def _stateTransfer(self, action: str) -> None:
+        try:
+            self._state = _model_automata[self._state][action]
+        except KeyError:
+            raise AutomataException(
+                f'{self._model_id}: Invalid action `{action}` in state `{self._state}`;\n  speed - {self._speed}m/s\n  cps - {self._checkpoints};\n  objs - {self._object_positions}')
+
+
+    def nearestObject(self, pos: float, after=None, speed=None, not_exact=False,
+                      preference='nearest') -> Optional[Tuple[Any, float]]:
+        if len(self._object_positions) == 0:
+            return None
+
+        if after is not None:
+            if speed is None:
+                speed = self._speed
+            objs = shift(self._object_positions, after * speed)
+        else:
+            objs = self._object_positions
+
+        res = search_pos(objs, pos, preference=preference)
+        if res is not None:
+            (oid, o_pos), idx = res
+            if not_exact and o_pos == pos:
+                if preference == 'prev':
+                    idx -= 1
+                elif preference == 'next':
+                    idx += 1
+                else:
+                    raise Exception('please dont use nearest with not exact')
+                if idx < 0 or idx > len(objs):
+                    return None
+                oid, o_pos = objs[idx]
+            return self._objects[oid], o_pos
+        return None
+
+    def putObject(self, obj_id: int, obj: Any, pos: float, soft_collide=True, return_nearest=False):
+        assert obj_id not in self._objects, "Clashing object ID!"
```

Comment:
Can you please clarify here assert message? For example, something like: `'There are multiple objects with the same ID!'`